### PR TITLE
fix doxygen warnings in cudf/io/types.hpp, other header files

### PR DESCRIPTION
**Pull request by karthikeyann**
Fixes parts of https://github.com/rapidsai/cudf/issues/9373
added missing documentation to fix doxygen warnings in multiple files

fixes 93 warnings.

### CODE REVIEW

File: cpp/include/cudf/io/datasource.hpp
Comment (vuule) [inline cpp/include/cudf/io/datasource.hpp:76]:
```suggestion
    /**
     * @brief Factory to construct a datasource buffer object from a container.
     *
     * @tparam Container Type of the container to construct the buffer from
     * @param data_owner The container to construct the buffer from (ownership is transferred)
     * @return Constructed buffer object
```

File: cpp/include/cudf/io/datasource.hpp
Comment (vuule) [inline cpp/include/cudf/io/datasource.hpp:321]:
```suggestion
     * @param data_owner The container to construct the buffer from (ownership is transferred)
```

File: cpp/include/cudf/io/datasource.hpp
Comment (vuule) [inline cpp/include/cudf/io/datasource.hpp:332]:
```suggestion
     * @param data_owner The container to construct the buffer from (ownership is transferred)
```

File: cpp/include/cudf/io/datasource.hpp
Comment (vuule) [inline cpp/include/cudf/io/datasource.hpp:370]:
```suggestion
 * @brief Factory to construct a datasource buffer object from a container.
 *
 * @tparam Container Type of the container to construct the buffer from
 * @param data_owner The container to construct the buffer from (ownership is transferred)
 * @return Constructed buffer object
```

File: cpp/include/cudf/io/datasource.hpp
Comment (vuule) [inline cpp/include/cudf/io/datasource.hpp:351]:
```suggestion
     * @brief Returns the pointer to the data in the buffer.
     *
     * @return Pointer to the data in the buffer
```

**Review by vuule — CHANGES_REQUESTED** [review]
Looks good for the most part. A few corrections/questions:

File: cpp/include/cudf/io/types.hpp
Comment (karthikeyann) [reply cpp/include/cudf/io/types.hpp:147]: This file had similar style at other places, so use this `//!<` for uniformity. No other reason.

File: cpp/include/cudf/io/datasource.hpp
Comment (karthikeyann) [reply cpp/include/cudf/io/datasource.hpp:370]: This docstring is copied from the declaration part of the code in same file. doxygen throws warning due to a limitation of template function.
shall we move the code definition to the declaration itself?

**Comment by ajschmidt8** [comment]
Removing `ops-codeowners` from the required reviews since it doesn't seem there are any file changes that we're responsible for. Feel free to add us back if necessary.

File: cpp/include/cudf/io/datasource.hpp
Comment (vuule) [reply cpp/include/cudf/io/datasource.hpp]: You can move the definition to the declaration, just need to forward declare `owning_buffer`. I think it's also possible to `copydoc` and avoid doc duplication that way.

Comment (karthikeyann) [reply cpp/include/cudf/io/datasource.hpp]: I tried `copydoc`, doxygen complains recursive reference.

Comment (karthikeyann) [reply cpp/include/cudf/io/datasource.hpp]: done.

**Review by vuule — APPROVED** [review]

File: cpp/include/cudf/io/types.hpp
Comment (vyasr) [inline cpp/include/cudf/io/types.hpp:582]:
```suggestion
 * @brief Metadata for a table
```

**Review by vyasr — APPROVED** [review]
One minor typo fix. Also, I asked this on another of your PRs: I'm not sure why `pure` is needed based on the contents of the doxygen docs for the pure command. I'm approving assuming you can address that question consistently across all the doxygen PRs. Thanks!

**Comment by karthikeyann** [comment]
@gpucibot merge

**Pull request merged**
